machine/v1/types_controlplanemachineset.go

```diff
@@ -284,6 +289,17 @@ type GCPFailureDomain struct {
 	Zone string `json:"zone"`
 }
 
+// OpenStackFailureDomain configures failure domain information for the OpenStack platform.
+type OpenStackFailureDomain struct {
```

Comment:
https://www.fir3net.com/Cloud/Openstack/openstack-partitioning-regions-availability-zones-and-host-aggregates.html mentions there are three OpenStack partitions: Regions, Availability Zones & Host Aggregates. Any chance host aggregates might be added later? Are there other failure domain information that might change the shape of `OpenStackFailureDomain`? `AWSFailureDomain` includes subnet part which moves zones under a new `AWSFailureDomainPlacement` type. Something that might be better to do right now to avoid growing a list of distinct information (e.g. placement, networking) on the same level.

Reply:
> Any chance host aggregates might be added later? Are there other failure domain information that might change the shape of `OpenStackFailureDomain`?

Yes, and they'll be added later but they'll always be optional and default to empty.

> `AWSFailureDomain` includes subnet part which moves zones under a new `AWSFailureDomainPlacement` type. Something that might be better to do right now to avoid growing a list of distinct information (e.g. placement, networking) on the same level.

I guess we could, we could have:

```golang
type OpenStackFailureDomain struct {
	// Placement configures the placement information for this instance.
	// +optional
	Placement AWSFailureDomainPlacement `json:"placement,omitempty"`

	// Storage is a reference to the storage to use for this instance.
	// +optional
	Storage *OpenStackFailureDomainStorage `json:"storage,omitempty"`

	// Network is a reference to the network to use for this instance.
	// +optional
	Network *OpenStackFailureDomainNetwork `json:"network,omitempty"`

}

type OpenStackFailureDomainPlacement struct {
	// AvailabilityZone is the availability zone of the instance.
	// +kubebuilder:validation:Required
	AvailabilityZone string `json:"availabilityZone"`
}

type OpenStackFailureDomainStorage struct {
	// AvailabilityZone is the availability zone of the volume.
	// +kubebuilder:validation:Required
	AvailabilityZone string `json:"availabilityZone"`

        // Later we could have other fields for Type, and additional volumes.
}

// Later we could have another type for Networking
type OpenStackFailureDomainNetwork struct {
	// Subnet(s) where the instance is attached to
}
```

WDYT?

Reply:
> [fir3net.com/Cloud/Openstack/openstack-partitioning-regions-availability-zones-and-host-aggregates.html](https://www.fir3net.com/Cloud/Openstack/openstack-partitioning-regions-availability-zones-and-host-aggregates.html) mentions there are three OpenStack partitions: Regions, Availability Zones & Host Aggregates. Any chance host aggregates might be added later?

From a nova (compute) perspective, availability zones are just a special type of host aggregate. A host can belong to multiple host aggregates and this information is only exposed to admins, not end-users, so I couldn't imagine anyone would ever use them for failure domain purposes. I'm not sure it ever makes sense to consider them.

Regions are purely a keystone (identity) thing. They allow you to share a single keystone instance across multiple OpenStack deployments. Because it's not possible to e.g. migrate instances between separate deployments, I don't think it would be reasonable to think of these as a failure domain either.

> Are there other failure domain information that might change the shape of `OpenStackFailureDomain`?

Based on the above, I would think not. We already cover compute, storage (both cinder and manila), and networking. This should be more than sufficient and is how any HA configuration I've seen has been handled.

> Besides already mentioned [volume type](https://github.com/openshift/api/pull/1442#issuecomment-1525584951). `AWSFailureDomain` includes subnet part which moves zones under a new `AWSFailureDomainPlacement` type. Something that might be better to do right now to avoid growing a list of distinct information (e.g. placement, networking) on the same level.

Reply:
In addition to what @stephenfin wrote, we have explicitly pushed Networking out of scope of our initial failure domains. We have several problems here:
* Neutron doesn't have a comprehensive AZ concept in the way Nova and Cinder do
* Unlike in public cloud, it is common in on-prem clusters to have multiple networks.
* There is no requirement for a user to segregate all possible network interfaces by failure domain
* The definition of a network interface is substantially more complex than simply what subnet it is connected to (e.g. hw offload, trunking, etc), which we don't want to add to the failure domain

Also, almost nobody is going to use it in OpenStack. The use case for networking like this is strict spine-leaf deployments and the one customer who was asking for this decided it wasn't important as they originally thought.

In short, it was hard and we punted it so we could land the rest.

Reply:
In AWS, we have the additional step in the API to add structure because this matches the `providerSpec`. The idea is that the failure domain itself represents a subset of the `providerSpec`, so to a user, it's obvious which fields from the failure domain map where in the provider spec. So I think we should keep the same pattern for this as well

Reply:
that makes sense Joel, I'll implement that.

Reply:
@JoelSpeed @ingvagabund I followed your advice, what do you think of this proposal?

Reply:
Looks better to me now yes, looks like it's closer to matching the providerSpec structure. One nit though, we don't use pointers in APIs in general (bar a few exceptions, eg Discriminated Unions), so, can we change the rootVolume to not be a pointer, will make coding against it easier.

Unless I've missed some reason that this _must_ be a pointer, but I can't think of one at the moment

Reply:
> // If not specified, no root volume will be created.
> // +optional
> RootVolume *RootVolume `json:"rootVolume,omitempty"`

Turning the field into non-pointer would change the semantics. I.g. `"rootVolume": {"availabilityZone": ""}` would mean "create a volume without the az specified (random)" as opposed to "do not create any root volume".

Reply:
I had thought we had a requirement in the failure domain that the AZ fields were not permitted to be empty, so I would expect if it were a non-pointer we would say
```
if failureDomain.RootVolume.AvailabilityZone != "" {
  // User has an opinion about the Az here, so lets make sure we set it on the providerSpec
  ...
}
```

Reply:
Perhaps a clarifying comment would be good for these fields, and perhaps a min-length to make sure when it is provided, it isn't empty?
```
When omitted, this means no opinion and the value will be taken from any existing value present in the machine template providerSpec.
```

Reply:
> I had thought we had a requirement in the failure domain that the AZ fields were not permitted to be empty

I see. Can it be always guaranteed? I mean `RootVolume` had two fields originally (zone and volume type). Now there's only a single one. In case another field gets introduced, will the `AvailabilityZone` field be always required to be non-empty? Or, will the requirement be to have at least one field in the `RootVolume` struct non-empty?

Reply:
What I am asking is: Is the `RootVolume` struct always going to have at least one AZ field set? Either the current `AvailabilityZone` field or maybe-to-be-added `VolumeZone` field? And if not (i.e. both fields are empty), will it mean "no root volume will be created"? Currently, the `RootVolume` struct has `+kubebuilder:validation:MinProperties:=1` which means `If not specified, no root volume will be created.` is not a valid case.

Reply:
> I had thought we had a requirement in the failure domain that the AZ fields were not permitted to be empty,

Is this requirement documented? If this is a universal requirement, the `AvailabilityZone` field under the `OpenStackFailureDomain` struct should be required as well.

Reply:
So I think this boils down to what is valid when we look at the structure
```
openstack:
- availabiltyZone: foo
- rootVolume:
    availabailityZone: foo
- availabilityZone: foo
  rootVolume:
    availabilityZone: foo
- rootVolume:
    someNewField: foo
```
IMO these are all valid
Whereas the following would not be valid:
```
openstack:
- {}
- rootVolume: {}
- availabilityZone: ""
- rootVolume:
    availabilityZone: ""
```

So I think this boils down to a few requirements:
* the failuredomain for openstack must have at least one, non-empty field
* the rootVolume, must have at least one non-empty field
* If specified, an availabilityZone must have length at least 1

Which makes me think that the fields for AZs should be optional, but using min-properties we can make the structure make sense

Perhaps some integration tests for these would help? 

Oh and, because this is all from the view of an unstructured client, whether this field is a pointer or not doesn't actually matter, what matters is the json serialization, so, do we have the right omitempty tags

Reply:
```yaml
- rootVolume:
    availabilityZone: ""
```
```
// +kubebuilder:validation:MinProperties:=1
type RootVolume struct {
...
	// +optional
	AvailabilityZone string `json:"availabilityZone,omitempty"`
```
Common sense tells me `kubebuilder:validation:MinProperties:=1` together with `AvailabilityZone` as optional field does not work well together. Though, there might be a bigger picture of the validation mechanism I don't see. 
From https://book.kubebuilder.io/reference/markers/crd-validation.html:
```
// +kubebuilder:validation:MinProperties int
restricts the number of keys in an object
```
Meaning one needs to have at least one field present or set? When preset, the value can be nil. So `{RootVolume: {AvailabilityZone: ""}}` should pass.

Anyway, I agree the unit tests need to validate the mentioned cases.

Reply:
@JoelSpeed  : 

> So I think this boils down to what is valid when we look at the structure
(...)

This is all covered by the tests; so we're good here.


> Whereas the following would not be valid:

Let me go line by line:

---
```
openstack:
- {}
```

A test was missing, I added one.

---
```
openstack:
- rootVolume: {}
```

This one is tested already:
```
  - name: Should reject an OpenStack failure domain with an empty rootVolume provided
    initial: |
      apiVersion: machine.openshift.io/v1
      kind: ControlPlaneMachineSet
      spec:
            [...]
            spec:
              providerSpec: {}
            failureDomains:
              platform: OpenStack
              openstack:
              - availabilityZone: foo
                rootVolume: {}
    expectedError: "spec.template.machines_v1beta1_machine_openshift_io.failureDomains.openstack[0].rootVolume in body should have at least 1 properties"
```

And it works as expected.

---
```
openstack:
- availabilityZone: ""
```

I added a test and I noticed that I had no error, so I added `+kubebuilder:validation:MinLength=1` to `OpenStackFailureDomain.AvailabilityZone`, and now the test fails as we expect.

---
```
openstack:
- rootVolume:
    availabilityZone: ""
```

I added a test and I noticed that I had no error, so I added `+kubebuilder:validation:MinLength=1` to `RootVolume.AvailabilityZone`, and now the test fails as we expect.

---

@ingvagabund :

> Common sense tells me kubebuilder:validation:MinProperties:=1 together with AvailabilityZone as optional field does not work well together. Though, there might be a bigger picture of the validation mechanism I don't see.

I'll let @JoelSpeed comment on that, but looking at the tests I think we now cover every possible scenario.

Reply:
```
// +kubebuilder:validation:MinLength=1
// +optional
AvailabilityZone string `json:"availabilityZone,omitempty"`
```
combination of `+optional` and `+kubebuilder:validation:MinLength=1` makes me wonder about the same. They seem to complement each other.

Reply:
After detailed analysis of the kube-openapi code the `+optional` has no additional effect as long as `omitempty` is used. 
- With only `omitempty` specified neither `Default: ""` nor `Required:` fields in the schema definition are generated
- With both `+optional` and `omitempty` specified, neither `Default: ""` nor `Required:` fields in the schema definition are generated
- With only `+optional` specified the schema generates additional `Default: ""` field in the schema definition.
- Without both `+optional` and `omitempty` specified,  I get additional `Required: []string{...}` field generated in the schema definition

Reply:
Back to:
```
// +kubebuilder:validation:MinLength=1
// +optional
AvailabilityZone string `json:"availabilityZone,omitempty"`
```
This no longer complements each other.
- `+optional` means `AvailabilityZone` is not required
- `+kubebuilder:validation:MinLength=1` applies only when `AvailabilityZone` field is specified. If it is its minLength is 1.

Back to:
```
// +kubebuilder:validation:MinProperties:=1
type RootVolume struct {
...
	// +optional
	AvailabilityZone string `json:"availabilityZone,omitempty"`
```
This together with `+kubebuilder:validation:MinLength=1` for the `AvailabilityZone` addresses the `{RootVolume: {AvailabilityZone: ""}}` case.

Reply:
Thanks @ingvagabund for this deep analysis.